packages/framework/ai-collab/src/explicit-strategy/typeGeneration.ts

```diff
@@ -0,0 +1,373 @@
+				explanation: z.string().describe(editDescription),
+				content: generateDomainTypes
+					? getType(schema.allowedTypes)
+					: z.any().describe("Domain-specific content here"),
+			})
+			.describe(
+				"Replaces the tree with a specific value. This is useful for initializing the tree or replacing the state entirely if appropriate.",
+			);
+		const insert = z
+			.object({
+				type: z.literal("insert"),
+				explanation: z.string().describe(editDescription),
+				content: generateDomainTypes
+					? getType(insertSet)
+					: z.any().describe("Domain-specific content here"),
+				destination: z.union([arrayPlace, objectPlace]),
+			})
+			.describe("Inserts a new object at a specific Place or ArrayPlace.");
+		const remove = z
+			.object({
+				type: z.literal("remove"),
+				explanation: z.string().describe(editDescription),
+				source: z.union([objectTarget, range]),
+			})
+			.describe("Deletes an object or Range of objects from the tree.");
+		const move = z
+			.object({
+				type: z.literal("move"),
+				explanation: z.string().describe(editDescription),
+				source: z.union([objectTarget, range]),
+				destination: z.union([arrayPlace, objectPlace]),
+			})
+			.describe("Moves an object or Range of objects to a new Place or ArrayPlace.");
+		const modify = z
+			.object({
+				type: z.enum(["modify"]),
+				explanation: z.string().describe(editDescription),
+				target: objectTarget,
+				field: z.enum([...modifyFieldSet] as [string, ...string[]]), // Modify with appropriate fields
+				modification: generateDomainTypes
+					? getType(modifyTypeSet)
+					: z.any().describe("Domain-specific content here"),
+			})
+			.describe("Sets a field on a specific ObjectTarget.");
+		const editTypes = [setRoot, insert, remove, move, modify, z.null()] as const;
+		const editWrapper = z.object({
+			edit: z
+				.union(editTypes)
+				.describe("The next edit to apply to the tree, or null if the task is complete."),
+		});
+		const typeRecord: Record<string, Zod.ZodTypeAny> = {
+			ObjectTarget: objectTarget,
+			ObjectPlace: objectPlace,
+			ArrayPlace: arrayPlace,
+			Range: range,
+			SetRoot: setRoot,
+			Insert: insert,
+			Remove: remove,
+			Move: move,
+			Modify: modify,
+			EditWrapper: editWrapper,
+		};
+		return [typeRecord, "EditWrapper"];
+	});
+}
+const editDescription =
+	"A description of what this edit is meant to accomplish in human readable English";
+function getOrCreateType(
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+	typeMap: Map<string, Zod.ZodTypeAny>,
+	insertSet: Set<string>,
+	modifyFieldSet: Set<string>,
+	modifyTypeSet: Set<string>,
+	definition: string,
+): Zod.ZodTypeAny {
+	return getOrCreate(typeMap, definition, () => {
+		const nodeSchema = definitionMap.get(definition) ?? fail("Unexpected definition");
+		switch (nodeSchema.kind) {
+			case NodeKind.Object: {
+				for (const [key, field] of Object.entries(nodeSchema.fields)) {
+					// TODO: Remove when AI better
```

Comment:
Can we say something more specific here? @taylorsw04 @noencke

Reply:
@noencke I remember we had to prevent modifications to fields that could be arrays, and I'm fairly sure it was because the model would clobber arrays rather than inserting/removing to accomplish a goal, but I'm not 100% on that. Can you validate?

Reply:
Unfortunately I don't remember, but what you said makes sense and seems the likely reason.